cpp/src/io/parquet/page_data.cu

```diff
@@ -1845,8 +1940,647 @@ __global__ void __launch_bounds__(block_size)
+                           uint32_t lane_id)
+{
+  using cudf::detail::warp_size;
+  // string_n is prefix(string_n-1) + suffix_n. in the worst case, copying
+  // a warp will take 32 iterations.
+  // exceptions:
+  // 1) all prefix_lens are 0. all data is in the suffix, already done.
+  // 2) all prefix_lens are equal. all threads copy last_string[0:prefix_len].
+  // 2a) all prefix_n-1 > prefix_n, can still go to last_string
+  // 2b) all prefix_lens are != 0, so there's a chance to copy the first
+  //     min(prefix_len) bytes from last_string, but this optimization probably
+  //     isn't worth the effort
+  //
+  // check for 1) and then loop
+  // for i = 0 -> 31:
+  //   check for 2) and return if possible
+  //   copy prefix_i bytes from string i-1 (last_string if i==0), update prefix_lens
+  //   check for cases where prefix_i < prefix_i+1, copy left->right, update prefix_lens
+  //   if all prefix_len == 0 break
+  // worst case for above is O(n) iterations, but hopefully will average out to O(log(n))
+  //
+  __shared__ __align__(8) uint8_t const* last_ptr;
+  __shared__ __align__(8) int64_t prefix_lens[warp_size];
+  __shared__ __align__(8) uint8_t const* offsets[warp_size];
+
+  uint32_t value_count    = dba->prefixes.value_count;
+  uint32_t const ln_idx   = start_idx + lane_id;
+  uint32_t const src_idx  = rolling_index(ln_idx);
+  uint64_t prefix_len     = ln_idx < value_count ? dba->prefixes.value[src_idx] : 0;
+  uint8_t* const lane_out = ln_idx < value_count ? strings_out + dba->offset[src_idx] : nullptr;
+  prefix_lens[lane_id]    = prefix_len;
+  offsets[lane_id]        = lane_out;
+
+  // if all prefix_len's are zero, then there's nothing to do
+  if (__all_sync(0xffff'ffff, prefix_len == 0)) { return; }
+
+  // initialize mask so lane 0 doesn't vote. will remove lane "i+1" at the end of each iteration
+  uint32_t mask = 0xffff'fffe;
+  if (lane_id == 0) { last_ptr = last_string; }
+  __syncwarp();
+
+  for (uint32_t i = 0; i < warp_size && i + start_idx < value_count; i++) {
+    // see if all prefixes are <= those of the neighbor to the left. if so, then copy prefix_len
+    // bytes from last_string and return.
+    if (__all_sync(mask, prefix_len <= prefix_lens[lane_id - 1])) {
+      if (lane_id >= i && lane_out != nullptr) { memcpy(lane_out, last_ptr, prefix_len); }
+      return;
+    }
+
+    if (prefix_lens[i] != 0) {
+      // current leader gets data from last_ptr
+      if (lane_id == i) {
+        memcpy(lane_out, last_ptr, prefix_len);
+        prefix_lens[lane_id] = prefix_len = 0;
+      }
+      __syncwarp();
+
+      if (lane_id > i) {
+        // copy data from the nearest prefix_len==0 neighbor to the left (if any),
+        // but do nothing if we hit a neighbor with a prefix_len < ours.
+        // this may be a bit of premature optimization, intended for cases like:
+        //  0  1  2  3  4
+        //  X  X  X  X  X
+        //  X     X     X
+        //  X     X
```

Comment:
Does the illustration represent prefix lengths [0, 2, 0, 2, 1]?

Reply:
Yes